drivers/include/drivers/interfaces/InterfaceCAN.h

```diff
@@ -0,0 +1,171 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License
+ */
+
+#ifndef MBED_INTERFACE_CAN_H_
+#define MBED_INTERFACE_CAN_H_
+
+#include "hal/can_helper.h"
+
+#include <cstring>
+
+#include "platform/Callback.h"
+
+namespace mbed {
+
+// Forward declare CAN
+class CAN;
+
+/** \defgroup drivers-public-api-can CAN
+ * \ingroup drivers-public-api
+ */
+
+/**
+ * \defgroup drivers_CANMessage CANMessage class
+ * \ingroup drivers-public-api-can
+ * @{
+ */
+
+/** CANMessage class
+ *
+ * @note Synchronization level: Thread safe
+ */
+class CANMessage : public CAN_Message {
+
+public:
+    /** Creates empty CAN message.
+     */
+    CANMessage() : CAN_Message()
+    {
+        len    = 8U;
+        type   = CANData;
+        format = CANStandard;
+        id     = 0U;
+        memset(data, 0, 8);
+    }
+
+    /** Creates CAN message with specific content.
+     *
+     *  @param _id      Message ID
+     *  @param _data    Mesaage Data
+     *  @param _len     Message Data length
+     *  @param _type    Type of Data: Use enum CANType for valid parameter values
+     *  @param _format  Data Format: Use enum CANFormat for valid parameter values
+     */
+    CANMessage(unsigned int _id, const unsigned char *_data, unsigned char _len = 8, CANType _type = CANData, CANFormat _format = CANStandard)
+    {
+        len    = (_len > 8) ? 8 : _len;
+        type   = _type;
+        format = _format;
+        id     = _id;
+        memcpy(data, _data, len);
+    }
+
+
+    /** Creates CAN message with specific content.
+     *
+     *  @param _id      Message ID
+     *  @param _data    Mesaage Data
+     *  @param _len     Message Data length
+     *  @param _type    Type of Data: Use enum CANType for valid parameter values
+     *  @param _format  Data Format: Use enum CANFormat for valid parameter values
+     */
+    CANMessage(unsigned int _id, const char *_data, unsigned char _len = 8, CANType _type = CANData, CANFormat _format = CANStandard)
+    {
+        len    = (_len > 8) ? 8 : _len;
+        type   = _type;
+        format = _format;
+        id     = _id;
+        memcpy(data, _data, len);
+    }
+
+    /** Creates CAN remote message.
+     *
+     *  @param _id      Message ID
+     *  @param _format  Data Format: Use enum CANType for valid parameter values
+     */
+    CANMessage(unsigned int _id, CANFormat _format = CANStandard)
+    {
+        len    = 0;
+        type   = CANRemote;
+        format = _format;
+        id     = _id;
+        memset(data, 0, 8);
+    }
+};
+
+/** @}*/
+
+namespace interface {
+
+namespace can {
```

Comment:
That won't play well with code using the enum values: `CAN::Reset` won't be found as a member for example. As a workaround, this can be wrapped inside a structure instead of a namespace and `interface::Can` can inherit from it .  If polymorphism is disabled then `mbed::Can` inherits from it too. 

```c++
struct can { 

enum Mode {
// enum definition
};

// Prevent slicing and user creation of base class.
protected:
    can() = default;
    ~can() = default;
    can(const can&) = default;
    can& operator=(const can&) = default;
};
```